systest/cluster/nodes.go

```diff
@@ -726,6 +733,66 @@ func deployNode(ctx *testcontext.Context, id string, labels map[string]string, f
 	for _, flag := range flags {
 		cmd = append(cmd, flag.Flag())
 	}
+
+	podSpec := corev1.PodSpec().
+		WithNodeSelector(ctx.NodeSelector).
+		WithVolumes(
+			corev1.Volume().WithName("config").
+				WithConfigMap(corev1.ConfigMapVolumeSource().WithName(spacemeshConfigMapName)),
+			corev1.Volume().WithName("data").
+				WithEmptyDir(corev1.EmptyDirVolumeSource().
+					WithSizeLimit(resource.MustParse(ctx.Storage.Size))),
+		).
+		WithDNSConfig(corev1.PodDNSConfig().WithOptions(
+			corev1.PodDNSConfigOption().WithName("timeout").WithValue("1"),
+			corev1.PodDNSConfigOption().WithName("attempts").WithValue("5"),
+		)).
+		WithContainers(corev1.Container().
+			WithName("smesher").
+			WithImage(ctx.Image).
+			WithImagePullPolicy(apiv1.PullIfNotPresent).
+			WithPorts(
+				corev1.ContainerPort().WithContainerPort(7513).WithName("p2p"),
+				corev1.ContainerPort().WithContainerPort(9092).WithName("grpc-pub"),
+				corev1.ContainerPort().WithContainerPort(9093).WithName("grpc-priv"),
+				corev1.ContainerPort().WithContainerPort(9094).WithName("grpc-post"),
+				corev1.ContainerPort().WithContainerPort(prometheusScrapePort).WithName("prometheus"),
+				corev1.ContainerPort().WithContainerPort(phlareScrapePort).WithName("pprof"),
+			).
+			WithVolumeMounts(
+				corev1.VolumeMount().WithName("data").WithMountPath("/data"),
+				corev1.VolumeMount().WithName("config").WithMountPath(configDir),
+			).
+			WithResources(corev1.ResourceRequirements().
+				WithRequests(smesherResources.Get(ctx.Parameters).Requests).
+				WithLimits(smesherResources.Get(ctx.Parameters).Limits),
+			).
+			WithStartupProbe(
+				corev1.Probe().WithTCPSocket(
+					corev1.TCPSocketAction().WithPort(intstr.FromInt32(9092)),
+				).WithInitialDelaySeconds(10).WithPeriodSeconds(10),
+			).
+			WithEnv(
+				corev1.EnvVar().WithName("GOMAXPROCS").WithValue("4"),
+			).
+			WithCommand(cmd...),
+		)
+
+	if key != nil {
+		podSpec = podSpec.
+			WithInitContainers(
+				corev1.Container().
+					WithName("file-creator").
+					WithImage("busybox").
+					WithCommand("sh", "-c",
+						fmt.Sprintf("mkdir -p /data/identities && echo '%x' > /data/identities/local.key", key),
+					).
```

Comment:
Although it's unlikely that `key` will break escaping, it might be a little bit better to pass it as a separate argument and avoid using `fmt.Sprintf` to construct the command:
```suggestion
					WithCommand("sh", "-c",
						"mkdir -p /data/identities && echo "$1" > /data/identities/local.key", "sh", key,
					).
```

(could skip second `"sh"` and use `$0` too, but that's considered to be less well-behaved generally)

Reply:
`key` is not a string. It's a sequence of raw bytes that needs to be written to `local.key` as hex encoded string.

I'm not quite understanding how the `fmt.Sprintf` approach would fail escaping and how your suggested approach solves the issue?

Reply:
Sorry, overlooked that it's `%x` not `%s`, my bad. No problem here